cilium/cmd/monitor.go

```diff
@@ -59,40 +71,103 @@ var (
 		SampleType:   bpf.PERF_SAMPLE_RAW,
 		WakeupEvents: 1,
 	}
+	eventTypeIdx = -1 // for integer comparison
+	eventType    = ""
+	eventTypes   = map[string]int{
+		"drop":    bpfdebug.MessageTypeDrop,
+		"debug":   bpfdebug.MessageTypeDebug,
+		"capture": bpfdebug.MessageTypeCapture,
+	}
+	fromSource = uint16(0)
+	toDst      = uint32(0)
+	related    = uint32(0)
 )
 
 func lostEvent(lost *bpf.PerfEventLost, cpu int) {
 	fmt.Printf("Lost %d events\n", lost.Lost)
 }
 
-func receiveEvent(msg *bpf.PerfEventSample, cpu int) {
-	prefix := fmt.Sprintf("CPU %02d:", cpu)
+// match checks if the event type, from endpoint and / or to endpoint match
+// when they are supplied. The either part of from and to endpoint depends on
+// related to, which can match on both.  If either one of them is less than or
+// equal to zero, then it is assumed user did not use them.
+func match(messageType int, src uint16, dst uint32) bool {
+	if eventTypeIdx > bpfdebug.MessageTypeUnspec && messageType != eventTypeIdx {
+		return false
+	} else if fromSource > 0 && fromSource != src {
+		return false
+	} else if toDst > 0 && toDst != dst {
+		return false
+	} else if related > 0 && uint16(related) != src && related != dst {
+		return false
+	}
 
-	data := msg.DataDirect()
-	if data[0] == bpfdebug.MessageTypeDrop {
-		dn := bpfdebug.DropNotify{}
-		if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dn); err != nil {
-			fmt.Printf("Error while parsing drop notification message: %s\n", err)
-		}
+	return true
+}
+
+func dropEvents(prefix string, data []byte) {
+	dn := bpfdebug.DropNotify{}
+
+	if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dn); err != nil {
+		fmt.Printf("Error while parsing drop notification message: %s\n", err)
+	}
+	if match(bpfdebug.MessageTypeDrop, dn.Source, dn.DstID) {
 		dn.Dump(dissect, data, prefix)
-	} else if data[0] == bpfdebug.MessageTypeDebug {
-		dm := bpfdebug.DebugMsg{}
-		if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dm); err != nil {
-			fmt.Printf("Error while parsing debug message: %s\n", err)
-		} else {
-			dm.Dump(data, prefix)
-		}
-	} else if data[0] == bpfdebug.MessageTypeCapture {
-		dc := bpfdebug.DebugCapture{}
-		if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dc); err != nil {
-			fmt.Printf("Error while parsing debug capture message: %s\n", err)
-		}
+	}
+}
+
+func debugEvents(prefix string, data []byte) {
+	dm := bpfdebug.DebugMsg{}
+
+	if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dm); err != nil {
+		fmt.Printf("Error while parsing debug message: %s\n", err)
+	}
+	if match(bpfdebug.MessageTypeDebug, dm.Source, dm.Arg1) {
```

Comment:
Passing `dm.Arg1` as destination ID is wrong. In most cases, that is not the destination. We can't support `--to` for debug messages. Let's just pass in`0` for now.